### PR TITLE
HLD for Save on Set

### PR DESCRIPTION
**Pull request by rlucus**
# Pull Requests

- https://github.com/sonic-net/sonic-buildimage/pull/18481
- https://github.com/sonic-net/sonic-buildimage/pull/19309
- https://github.com/sonic-net/sonic-gnmi/pull/108
- https://github.com/sonic-net/sonic-gnmi/pull/242

### CODE REVIEW

**Review by tomek-US — DISMISSED** [review]
LGTM

**Review by anand-kumar-subramanian — DISMISSED** [review]

**Comment by zhangyanzhao** [comment]
community review recording https://zoom.us/rec/share/fhMIOkkDYx3INSShyKdXw9dXl8HqhBqQLxzG1mEUyvJrxdYbdGhofRS9Xs4MCh2g.aBeFosVoWBv3VBXi

File: doc/mgmt/gnmi/save_on_set.md [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+No effect on warm/fast boot
+
+### Restrictions/Limitations
+
+### Testing Requirements/Design
+
+#### Unit Test cases
+
+- No behavior change if the feature is not enabled.
+- A gNMI.Set() call should generate a signal to save ConfigDB.
```
Comment (venkatmahalingam) [inline doc/mgmt/gnmi/save_on_set.md:215]: Are we invoking "write memory" click call for every gNMI set operation when the save-on-set is enabled? not sure signalling is the word usage, please clarify.

File: doc/mgmt/gnmi/save_on_set.md [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+- Is this change specific to any platform? Are there dependencies for platforms to implement anything to make this feature work? If yes, explain in detail and inform community in advance.
+  - Any Platform
+- SAI API requirements, CLI requirements, ConfigDB requirements. Design is covered in following sections.
+
+#### Required Changes
+
+##### Telemetry Executable
+A new command-line parameter will be added to control the behavior of the save-on-set functionality. By default, i.e. when the option is not specified, the gNMI server will behave as it did before this change - the configuration will not be saved to a file without explicit action from the administrator for example by execution of a command via ssh connection.
+
+The new parameter will be: --with-save-on-set and when present it will configure a function pointer variable gnmi.SaveOnSet to point to a function that actually performs the save operation.
```
Comment (venkatmahalingam) [inline doc/mgmt/gnmi/save_on_set.md:111]: Are we getting this feature enable/disable from config_db field?

File: doc/mgmt/gnmi/save_on_set.md (hunk 1 above)
Comment (rlucus) [reply doc/mgmt/gnmi/save_on_set.md:215]: Reworded

File: doc/mgmt/gnmi/save_on_set.md (hunk 2 above)
Comment (rlucus) [reply doc/mgmt/gnmi/save_on_set.md:111]: Added clarification that the startup script will read from the DB to pass the value along to the command line argument on startup.

**Review by anand-kumar-subramanian — APPROVED** [review]

File: doc/mgmt/gnmi/save_on_set.md [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+### Definitions/Abbreviations
+
+- [gNMI](https://github.com/openconfig/reference/blob/master/rpc/gnmi/gnmi-specification.md) - Google Network Management Interface
+- UMF - Unified Management Framework
+
+### Overview
+
+Having configuration be persistant across switch reboot is a useful feature that is not currently implemented by UMF.
+
+The required behaviour is to save the configuration to a file that is used to populate the configuration database during the startup process every time the gNMI.Set() RPC call is performed.
```
Comment (venkatmahalingam) [inline doc/mgmt/gnmi/save_on_set.md:52]: Why do we need to save the config for every gNMI set operation? Is it possible to periodically save (may be 1 min window) if there is a change in the config from gNMI server?

Comment (rlucus) [reply doc/mgmt/gnmi/save_on_set.md:52]: Set is a low frequency operation and delaying the save-on-set would reduce accuracy of saved config if a reset were to happen.

**Comment by zhangyanzhao** [comment]
code PR is missed and can you please add @rlucus ?

**Review by venkatmahalingam — APPROVED** [review]

**Comment by zhangyanzhao** [comment]
code PR is not ready, move to backlog for future release

**Review by bhagatgit — APPROVED** [review]

**Comment by qiluo-msft** [comment]
Could you add the code PRs in a markdown table in this HLD PR's description?

File: doc/mgmt/gnmi/save_on_set.md [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+// SaveOnSet point to a function that is called to save changes of configuration
+// to a file. By default it points to an empty function - the configuration is not
+// saved to a file.
+var SaveOnSet = SaveOnSetDisabled
+
+func (s *Server) Set(ctx context.Context, req *gnmipb.SetRequest) (*gnmipb.SetResponse, error) {
+
+  // ...
+
+  SaveOnSet()
```
Comment (ganglyu) [inline doc/mgmt/gnmi/save_on_set.md:149]: If set operations failed here, do you still need SaveOnSet?

File: doc/mgmt/gnmi/save_on_set.md [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+  - Any Platform
+- SAI API requirements, CLI requirements, ConfigDB requirements. Design is covered in following sections.
+
+#### Required Changes
+
+##### Telemetry Executable
+
+A new command-line parameter will be added to control the behavior of the save-on-set functionality. By default, i.e. when the option is not specified, the gNMI server will behave as it did before this change - the configuration will not be saved to a file without explicit action from the administrator for example by execution of a command via ssh connection.
+
+The telemetry.sh startup script will be modified to read `TELEMETRY|gnmi|save_on_set` and pass the value along during startup. Any changes to this value will only take effect after restarting telemetry.
```
Comment (isabelmsft) [inline doc/mgmt/gnmi/save_on_set.md:112]: Does this indicate that save_on_set is added as a field to the TELEMETRY table in ConfigDB? If so, we need to update yang model (and update CLI/YANG model requirements section of this HLD)

File: doc/mgmt/gnmi/save_on_set.md (hunk 4 above)
Comment (rlucus) [reply doc/mgmt/gnmi/save_on_set.md:149]: The call to save on set is the last thing before returning. If there are errors set failures the function returns the err early and the save is never called.

Looking over the code I do see one path were a failure is possible and the code doesn't return early. We could make this an explicit check if its necessary.

File: doc/mgmt/gnmi/save_on_set.md [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+| v0.2 | 03/26/2024 | Ryan Lucus (Google) | Update to use Sonic Service Client |
+
+### Scope
+
+Add the option for enabling the SONiC gNMI implementation to save its in-memory stored configuration to a file every time it changes.
+
+### Definitions/Abbreviations
+
+- [gNMI](https://github.com/openconfig/reference/blob/master/rpc/gnmi/gnmi-specification.md) - Google Network Management Interface
+- UMF - Unified Management Framework
```
[the remark this answers is not in the document]
Comment (rlucus) [reply doc/mgmt/gnmi/save_on_set.md:47]: UMF was the name for sonic-gnmi/sonic-mgmt-common
I'm not aware of any current references to it. Maybe just called Sonic-Gnmi now?

File: doc/mgmt/gnmi/save_on_set.md (hunk 5 above)
Comment (rlucus) [reply doc/mgmt/gnmi/save_on_set.md:112]: This is not intended to change the yang model. It is only read or written by the telemetry container to enable persistence in the feature.

**Comment by rlucus** [comment]
> Could you add the code PRs in a markdown table in this HLD PR's description?

# Pull Requests

- https://github.com/sonic-net/sonic-buildimage/pull/18481
- https://github.com/sonic-net/sonic-buildimage/pull/19309
- https://github.com/sonic-net/sonic-gnmi/pull/108
- https://github.com/sonic-net/sonic-gnmi/pull/242

File: doc/mgmt/gnmi/save_on_set.md (hunk 5 above)
Comment (rlucus) [reply doc/mgmt/gnmi/save_on_set.md:113]: https://github.com/sonic-net/sonic-buildimage/pull/19309

File: doc/mgmt/gnmi/save_on_set.md (hunk 4 above)
Comment (rlucus) [reply doc/mgmt/gnmi/save_on_set.md]: https://github.com/sonic-net/sonic-gnmi/pull/242

File: doc/mgmt/gnmi/save_on_set.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (rlucus) [reply doc/mgmt/gnmi/save_on_set.md]: changed.

**Review by bhagatgit — APPROVED** [review]

**Comment by bhagatgit** [comment]
Approving/merging this HLD as the code PRs are already submitted.

**Pull request merged**